Fix sending tick message not at exact tick.

ZehM4tt opened the pull request:
For unknown reasons the tick message was further delayed for another 25ms causing clients having the need to wait longer than required. Since network_update is called per Tick this should send the message at the given time and not further delay it, ticks already run per 25ms.

janisozaur commented:
> For unknown reasons

I don't like that this is unknown. Isn't there a code higher up the stack ensuring this has to be at least 25? Can you try tracking it down?

ZehM4tt commented:
It might be a left over since I refactored and moved quite a few things in the past PRs of the network code. game_update is called per Tick which remains to be 25ms. There is really no reason to add any delay in sending the tick message.

janisozaur commented:
This is a good candidate for 0.1.0 still, but needs review.

zsilencer commented:
The reason is because network_update is not called per tick when in uncap fps mode, so if your screen is 140fps it would be sending out 140 packets per second to each client with this.  The further delay may actually be just that it is not synchronized with _lastTick.

ZehM4tt commented:
That is not happening, it would mean game_logic_update runs 140 times per. I've recorded the delta between each Server_Send_Tick and its between 25 and 33 with the PR that fixes the accuracy. Its all how it should be.

zsilencer commented:
Looks like you changed it, but it will still send out same rate of packets as fps when paused.  Also, if the server changes game speed it will blast out tons of packets, that's why this check is here.  The network is supposed to be decoupled from the actual game running.  Pretty sure that the jitter is caused by _lastTick and last_tick_sent_time not matching up but haven't tested.

ZehM4tt commented:
I think I already have found the actual culprit, we should still find a way to send the message per tick just not when its paused. As for the game speed, maybe we should have the network rates/speed being a different thing.

ZehM4tt commented:
So I did a few more tests, turns out its actually fine to keep it. This was fixed by a semi-fixed timestep #5848  . This PR is not required anymore.
